File: src/main/scala/category/array/RandomizedSet.scala
package category.array

import java.util.Random

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

class RandomizedSet() {
  
  /** Initialize your data structure here. */
  val data                    = new mutable.HashSet[Int]
  var array: ArrayBuffer[Int] = new ArrayBuffer[Int]()
  val ran                     = new Random()
  
  
  /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
  def insert(`val`: Int): Boolean = {
    if (data.contains(`val`)) return false
    array.append(`val`)
    data.add(`val`)
  }
  
  /** Removes a value from the set. Returns true if the set contained the specified element. */
  def remove(`val`: Int): Boolean = {
    if (!data.contains(`val`)) return false
    array.remove(removeNumIndex(`val`))
    data.remove(`val`)
  }
  
  /** Get a random element from the set. */
  def getRandom(): Int = {
    array(ran.nextInt(data.size))
  }
  
  def removeNumIndex(num: Int): Int = {
    array.foldLeft(0) { case (idx, inner) =>
      if (inner == num) return idx
      idx + 1
    }
  }
  
}

/**
  * Your RandomizedSet object will be instantiated and called as such:
  * var obj = new RandomizedSet()
  * var param_1 = obj.insert(`val`)
  * var param_2 = obj.remove(`val`)
  * var param_3 = obj.getRandom()
  */

File: src/main/python/RandomizedSet.py
import random


class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.dict = {}
        self.lst = []

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.dict: return False
        self.dict[val] = len(self.lst)
        self.lst.append(val)
        return True

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val not in self.dict or len(self.dict) == 0: return False
        last, idx = self.lst[-1], self.dict[val]
        self.lst[idx], self.dict[last] = last, idx
        del self.dict[val]
        self.lst.pop()
        return True

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        if len(self.lst) == 0: return -1
        return random.choice(self.lst)

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
